output.py: support title setting via curses/terminfo

gerstner-hub opened the pull request:
To support terminal title setting for a broader range of terminals, implement curses/terminfo support for setting terminal titles, instead of relying only on the hard coded list of supported $TERM values.

gerstner-hub committed:
output.py: try to use curses for setting xterm title

Currently only the list of hard coded terminal names found in
_legal_terms_re allow usage of the title setting feature. Maintaining
this list is a burden and means that some terminals will miss out this
feature, although they actually do have support for it.

Thus, try to use the curses module to dynamically query the terminfo
database for "tsl" (to status line) and "fsl" (from status line)
strings. This way all terminals that declare support for the feature
will automatically benefit.

As a fallback, should curses not be available, still rely on the
_legal_terms_re for setting up terminal titles.

Signed-off-by: Matthias Gerstner <matthias.gerstner@nefkom.net>

gerstner-hub committed:
output.py: harmonize variable names and group xterm title code

The global variables are inconsistently named and spread over the
module, similarly the title related code is not grouped together.
Harmonize this.

Also the double negative logic involving `_disable_xtermTitle` makes the
code harder to read. Invert the logic.

Signed-off-by: Matthias Gerstner <matthias.gerstner@nefkom.net>

zmedico commented on lib/portage/output.py:
```diff
@@ -268,26 +268,67 @@ def nc_len(mystr):
+    try:
+        # by default check if we can dynamically query the proper title
+        # setting sequence via terminfo
+        curses = init_curses(sys.stderr)
```
We should be consistent in our usage of `sys.__stderr__` vs. `sys.stderr`, and `sys.__stderr__` is preferable because sometimes we override `sys.stderr` with a temporary `io.StringIO` instance.

zmedico replied:
Maybe it's better if the fd argument defaults to `sys.__stderr__` in init_curses, so we'll only have to explicitly pass it in for an unusual case.

gerstner-hub replied:
Right, I've pushed a commit on top to address this.

gerstner-hub committed:
output.py: consistently use sys.__stderr__

sys.stderr is sometimes temporarily replaced by other objects, thus rely
on sys.__stdserr__ consistently. Also provide a default for `fd` in
`init_curses()`.

zmedico reviewed (approved):
LGTM, thanks!
